File: 33_Search_in_a_rotated_array.cpp
// 33. Search in Rotated Sorted Array
// https://leetcode.com/problems/search-in-rotated-sorted-array/

class Solution {
public:
    int search(std::vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target)
                return mid;

            if (nums[mid] >= nums[left]) {
                if (target >= nums[left] && target < nums[mid])
                    right = mid - 1;
                else
                    left = mid + 1;
            } else {
                if (target > nums[mid] && target <= nums[right])
                    left = mid + 1;
                else
                    right = mid - 1;
            }
        }

        return -1;
    }
};

File: 33_Search_in_a_rotated_array.js
// 33. Search in Rotated Sorted Array
// https://leetcode.com/problems/search-in-rotated-sorted-array/
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let left = 0; right = nums.length-1;
    result = -1;

    while(left <= right){
        mid = Math.floor((left+right)/2);
        if(nums[mid] == target){
            result = mid;
        }
       
        if(nums[mid] >= nums[left]){
            if(target >= nums[left] && target <= nums[mid]){
                right = mid - 1;
            }
            else{
                left = mid+1;
            }
        }
        else{
            if(target >= nums[mid] && target <= nums[right]){
                left = mid+1;
            }
            else{
                right = mid-1;
            }
        }
    }
    return result;
}; 